synthesizer/src/vm/mod.rs

```diff
@@ -145,24 +143,25 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                     };
 
                     // Add the deployment and its imports found recursively.
-                    deployments.extend_from_slice(&load_deployment_and_imports(
-                        process,
-                        transaction_store,
-                        transaction_id,
-                    )?);
+                    load_deployment_and_imports(process, transaction_store, transaction_id, deployments.clone())?;
                 }
             }
 
             // Once all the imports have been included, add the parent deployment.
-            deployments.push((*program_id, deployment));
+            if !deployments.read().contains_key(program_id) {
+                // Note: There may be re-insertions due to parallelism, but it is safe to
+                //  reinsert into the IndexMap because it should be the same object.
+                deployments.write().entry(*program_id).or_insert(deployment);
+            }
```

Comment:
If we don't expect a lot of misses (i.e. duplicate entries), keeping only the inner expression would be faster.